File: android/src/main/java/com/whiteguru/capacitor/plugin/videoeditor/VideoEditorPlugin.java
package com.whiteguru.capacitor.plugin.videoeditor;

import android.Manifest;
import android.content.Context;
import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.getcapacitor.JSObject;
import com.getcapacitor.Logger;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

import android.util.Log;

import com.linkedin.android.litr.analytics.TrackTransformationInfo;
import com.linkedin.android.litr.TransformationListener;

import java.io.File;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.List;

@CapacitorPlugin(
        name = "VideoEditor",
        permissions = {
                @Permission(
                        strings = {Manifest.permission.READ_EXTERNAL_STORAGE},
                        alias = VideoEditorPlugin.STORAGE
                ),
                @Permission(
                        strings = {Manifest.permission.READ_MEDIA_VIDEO},
                        alias = VideoEditorPlugin.MEDIA_VIDEO
                )

        }
)
public class VideoEditorPlugin extends Plugin {

    // Permission alias constants
    static final String STORAGE = "storage";
    static final String MEDIA_VIDEO = "media_video";

    // Message constants
    private static final String PERMISSION_DENIED_ERROR_STORAGE = "User denied access to storage";

    @PluginMethod
    public void edit(PluginCall call) {
        editLitr(call);
    }

    public void editLitr(PluginCall call) {
        String path = call.getString("path");
        JSObject trim = call.getObject("trim", new JSObject());
        JSObject transcode = call.getObject("transcode", new JSObject());

        if (path == null) {
            call.reject("Input file path is required");
            return;
        }

        if (checkStoragePermissions(call)) {
            Uri inputUri = Uri.parse(path);
            File inputFile = new File(inputUri.getPath());

            if (!inputFile.canRead()) {
                call.reject("Cannot read input file: " + inputFile.getAbsolutePath());
                return;
            }

            String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.ENGLISH).format(new Date());
            String fileName = "VID_" + timeStamp + "_";
            File storageDir = getContext().getCacheDir();

            execute(
                    () -> {
                        try {
                            File outputFile = File.createTempFile(fileName, ".mp4", storageDir);

                            VideoEditorLitr implementation = new VideoEditorLitr();

                            TrimSettings trimSettings = new TrimSettings(trim.getInteger("startsAt", 0), trim.getInteger("endsAt", 0));

                            TranscodeSettings transcodeSettings = new TranscodeSettings(
                                    transcode.getInteger("height", 0),
                                    transcode.getInteger("width", 0),
                                    transcode.getBoolean("keepAspectRatio", true)
                            );

                            TransformationListener videoTransformationListener = new TransformationListener() {
                                @Override
                                public void onStarted(@NonNull String id) {
                                    Logger.debug("Transcode started");
                                }

                                @Override
                                public void onProgress(@NonNull String id, float progress) {
                                    Logger.debug("Transcode running " + progress);

                                    JSObject ret = new JSObject();
                                    ret.put("progress", progress);

                                    notifyListeners("transcodeProgress", ret);
                                }

                                @Override
                                public void onCompleted(@NonNull String id, @Nullable List<TrackTransformationInfo> trackTransformationInfos) {
                                    Logger.debug("Transcode completed");

                                    JSObject ret = new JSObject();
                                    ret.put("file", createMediaFile(outputFile));
                                    call.resolve(ret);
                                }

                                @Override
                                public void onCancelled(@NonNull String id, @Nullable List<TrackTransformationInfo> trackTransformationInfos) {
                                    Logger.debug("Transcode cancelled");

                                    call.reject("Transcode canceled");
                                }

                                @Override
                                public void onError(@NonNull String id, @Nullable Throwable cause, @Nullable List<TrackTransformationInfo> trackTransformationInfos) {
                                    Logger.debug("Transcode error: " + (cause != null ? cause.getMessage() : ""));

                                    call.reject("Transcode failed: " + (cause != null ? cause.getMessage() : ""));
                                }
                            };

                            implementation.edit(getContext(), inputFile, outputFile, trimSettings, transcodeSettings, videoTransformationListener);
                        } catch (Exception e) {
                            call.reject(e.getMessage());
                        }
                    }
            );
        }
    }

    @PluginMethod
    public void thumbnail(PluginCall call) {
        String path = call.getString("path");
        int atMs = call.getInt("at", 0);
        int width = call.getInt("width", 0);
        int height = call.getInt("height", 0);

        if (path == null) {
            call.reject("Input file path is required");
            return;
        }

        if (checkStoragePermissions(call)) {
            Uri inputUri = Uri.parse(path);

            String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.ENGLISH).format(new Date());
            String fileName = "TH_" + timeStamp + "_";
            File storageDir = getContext().getCacheDir();

            File outputFile = null;

            try {
                outputFile = File.createTempFile(fileName, ".jpg", storageDir);

                VideoEditorLitr implementation = new VideoEditorLitr();

                implementation.thumbnail(this.getContext(), inputUri, outputFile, atMs, width, height);
            } catch (Exception e) {
                call.reject(e.getMessage());
                return;
            }

            JSObject ret = new JSObject();
            ret.put("file", createMediaFile(outputFile));
            call.resolve(ret);
        }
    }

    private boolean checkStoragePermissions(PluginCall call) {
        if (getPermissionState(STORAGE) != PermissionState.GRANTED) {
            requestPermissionForAlias(STORAGE, call, "storagePermissionsCallback");
            return false;
        }
        return true;
    }

    /**
     * Completes the plugin call after a storage permission request
     *
     * @param call the plugin call
     */
    @PermissionCallback
    private void storagePermissionsCallback(PluginCall call) {
        Log.d("TAG", "Swag and then the getPermissionState(STORAGE) result: " + getPermissionState(STORAGE));
        if (getPermissionState(STORAGE) != PermissionState.GRANTED) {
            Logger.debug(getLogTag(), "User denied photos permission: " + getPermissionState(STORAGE).toString());
            call.reject(PERMISSION_DENIED_ERROR_STORAGE);
            return;
        }

        switch (call.getMethodName()) {
            case "edit":
                edit(call);
                break;
            case "thumbnail":
                thumbnail(call);
                break;
        }
    }

    /**
     * Creates a JSObject that represents a File from the Uri
     *
     * @param file the File of the audio/image/video
     * @return a JSObject that represents a File
     */
    private JSObject createMediaFile(File file) {
        Context context = getBridge().getActivity().getApplicationContext();
        Uri uri = Uri.fromFile(file);
        String mimeType = context.getContentResolver().getType(uri);

        JSObject ret = new JSObject();

        ret.put("name", file.getName());
        ret.put("path", uri);
        ret.put("type", mimeType);
        ret.put("size", file.length());

        return ret;
    }
}

File: ios/Plugin/VideoEditorPlugin.swift
import Foundation
import Capacitor
import MobileCoreServices

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(VideoEditorPlugin)
public class VideoEditorPlugin: CAPPlugin {
    private let implementation = VideoEditor()
    
    @objc func edit(_ call: CAPPluginCall) {
        let path = call.getString("path")?.replacingOccurrences(of: "file://", with: "");
        let trim = call.getObject("trim") ?? JSObject();
        let transcode = call.getObject("transcode") ?? JSObject();
        
        if (path == nil) {
            call.reject("Input file path is required");
            return;
        }
        
        do {
            let trimSettings = try TrimSettings(
                startsAt: (trim["startsAt"] ?? 0) as! CLong,
                endsAt: (trim["endsAt"] ?? 0) as! CLong
            );
            
            let transcodeSettings = try TranscodeSettings(
                height: (transcode["height"] ?? 0) as! Int,
                width: (transcode["width"] ?? 0) as! Int,
                keepAspectRatio: (transcode["keepAspectRatio"] ?? true) as! Bool
            );
            
            let outFile = self.getDestVideoUrl();
            
            DispatchQueue.main.async {
                self.implementation.edit(
                    srcFile: URL(fileURLWithPath: path!),
                    outFile: outFile,
                    trimSettings: trimSettings,
                    transcodeSettings: transcodeSettings,
                    completionHandler: { url in
                        call.resolve([
                            "file": self.createMediaFile(url: url),
                        ])
                    },
                    progressHandler: { [weak self] progress in
                        // User progress
                        self!.notifyListeners("transcodeProgress", data: ["progress": progress])
                    },
                    errorHandler: { error in
                        call.reject(error)
                    }
                );
            }
        } catch {
            call.reject("Invalid parameters")
        }
    }
    
    @objc func thumbnail(_ call: CAPPluginCall) {
        let path = call.getString("path");
        let atMs = call.getInt("at") ?? 0;
        let width = call.getInt("width") ?? 0;
        let height = call.getInt("height") ??  0;
        
        if (path == nil) {
            call.reject("Input file path is required");
            return;
        }
        
        let outFile = self.getDestImageUrl();
        
        do {
            try self.implementation.thumbnail(
                srcFile: URL(fileURLWithPath: path!),
                outFile: outFile,
                atMs: atMs,
                width: width,
                height: height
            );
            
            call.resolve([
                "file": self.createMediaFile(url: outFile),
            ]);
        } catch {
            call.reject("Invalid parameters")
        }
    }
    
    func getDestVideoUrl() -> URL {
        return URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent(NSUUID().uuidString)
            .appendingPathExtension("mp4")
    }
    
    func getDestImageUrl() -> URL {
        return URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent(NSUUID().uuidString)
            .appendingPathExtension("jpg")
    }
    
    func createMediaFile(url: URL) -> JSObject {
        let mimeType = self.getMimeType(url: url)
        var fileSize = 0
        
        do {
            let resources = try url.resourceValues(forKeys: [.fileSizeKey])
            fileSize = resources.fileSize!;
        } catch {
            //
        }
        
        var file = JSObject()
        
        file["name"] = url.lastPathComponent
        file["path"] = url.absoluteString
        file["type"] = mimeType
        file["size"] = fileSize
        
        return file;
    }
    
    func getMimeType(url: URL) -> String {
        let pathExtension = url.pathExtension
        if let uti = UTTypeCreatePreferredIdentifierForTag(kUTTagClassFilenameExtension, pathExtension as NSString, nil)?.takeRetainedValue() {
            if let mimetype = UTTypeCopyPreferredTagWithClass(uti, kUTTagClassMIMEType)?.takeRetainedValue() {
                return mimetype as String
            }
        }
        return "application/octet-stream"
    }
}
